nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java

```diff
@@ -0,0 +1,276 @@
+    }
+
+    @Override
+    public void clear() {
+        if (log.isTraceEnabled()) {
+            log.trace("MongoDB: CLEAR");
+        }
+        coll.deleteMany(new Document());
+    }
+
+    // if possible, store the bytes as a UTF-8 string
+    protected static Object toStorage(byte[] bytes) {
+        try {
+            return UTF_8_DECODERS.get().decode(ByteBuffer.wrap(bytes)).toString();
+        } catch (CharacterCodingException e) {
+            // could not decode as UTF-8, use a binary
+            return new Binary(bytes);
+        }
+    }
+
+    protected static byte[] fromStorage(Object value) {
+        if (value instanceof Binary) {
+            return ((Binary) value).getData();
+        } else if (value instanceof String) {
+            return ((String) value).getBytes(UTF_8);
+        } else {
+            throw new UnsupportedOperationException(value.getClass().getName());
+        }
+    }
+
+    @Override
+    public byte[] get(String key) {
+        Bson filter = eq(ID_KEY, key);
+        Document doc = coll.find(filter).first();
+        if (doc == null) {
+            if (log.isTraceEnabled()) {
+                log.trace("MongoDB: GET " + key + " = null");
+            }
+            return null;
+        }
+        Object value = doc.get(VALUE_KEY);
+        if (log.isTraceEnabled()) {
+            log.trace("MongoDB: GET " + key + " = " + value);
+        }
+        return fromStorage(value);
+    }
+
+    protected Date getDateFromTTL(long ttl) {
+        return new Date(System.currentTimeMillis() + ttl * 1000);
+    }
+
+    @Override
+    public void put(String key, byte[] bytes, long ttl) {
+        Bson filter = eq(ID_KEY, key);
+        if (bytes == null) {
+            if (log.isTraceEnabled()) {
+                log.trace("MongoDB: DEL " + key);
+            }
+            coll.deleteOne(filter);
+        } else {
+            Object value = toStorage(bytes);
+            Document doc = new Document(VALUE_KEY, value);
+            if (ttl != 0) {
+                doc.append(TTL_KEY, getDateFromTTL(ttl));
+            }
+            if (log.isTraceEnabled()) {
+                log.trace("MongoDB: PUT " + key + " = " + value + (ttl == 0 ? "" : " (TTL " + ttl + ")"));
+            }
+            coll.replaceOne(filter, doc, new UpdateOptions().upsert(true));
+        }
+    }
+
+    @Override
+    public boolean setTTL(String key, long ttl) {
+        Bson filter = eq(ID_KEY, key);
+        Bson update;
+        if (ttl == 0) {
+            update = unset(TTL_KEY);
+        } else {
+            update = set(TTL_KEY, getDateFromTTL(ttl));
+        }
+        if (log.isTraceEnabled()) {
+            log.trace("MongoDB: SETTTL " + key + " = " + ttl);
+        }
+        UpdateResult res = coll.updateOne(filter, update);
+        return res.getModifiedCount() == 1;
+    }
+
+    @Override
+    public boolean compareAndSet(String key, byte[] expected, byte[] value) {
+        Bson filter = eq(ID_KEY, key);
```

Comment:
here some comment will be welcome to explain how the concurrency is handled, also the method is too long

Reply:
Well concurrency and atomicity is handled at the database level by MongoDB itself, that's the goal.

Do you really think that the method is too long? It's 4 distinct cases with lots of logging and no complex logic.

Reply:
ok the compare and set is done with a replaceOne which is atomic but as it comes after 50 lines I simply didn't see it :)